src/digraph-cyclicity.jl

```diff
@@ -0,0 +1,371 @@
+            in(v, vis.blockedmap[w]) || push!(vis.blockedmap[w], v)
+        end
+    end
+    pop!(vis.stack)
+    return done
+end
+
+
+"""
+```simplecycles(dg::DiGraph)```
+
+Compute all cycles of the given directed graph, using 
+[Johnson's algorithm](http://epubs.siam.org/doi/abs/10.1137/0204007). 
+
+/!\ The number of cycles grow more than exponentially with the number of vertices, 
+you might want to use the algorithm with a ceiling -- `getcycles` -- on large directed graphs 
+(slightly slower). If you want to have an idea of the possible number of cycles, 
+look at function ```maxcycles(dg::DiGraph, byscc::Bool = true)```.
+
+# Arguments
+* `dg`: the directed graph
+
+# Returns
+* `cycles`: all the cycles of the directed graph
+"""
+function simplecycles(dg::DiGraph)
+    sccs = strongly_connected_components(dg)
+    cycles = Vector{Vector{Int}}()
+    for scc in sccs
+        while length(scc) > 1
+            wdg, vmap = induced_subgraph(dg, scc)
+            #startnode = 1
+            shift!(scc)
+            visitor = JohnsonVisitor(wdg)
+            circuit(1, wdg, visitor, cycles, vmap)
+        end
+    end
+    return cycles
+end
+
+
+##########################################################
+#### Iterative version, using Tasks, of the previous algorithms.
+"""
+```circuit(v::T, dg::DiGraph, vis::JohnsonVisitor, vmap::Vector{Int}, startnode = v)```
+
+One step of the recursive version of simple cycle detection, using a DFS algorithm.
+
+The CIRCUIT function from [Johnson's algorithm](http://epubs.siam.org/doi/abs/10.1137/0204007),
+ recursive and iterative version. Produce a cycle when needed, can be used only inside a 
+ Task.
+
+# Arguments
+* v: the vertex considered in this iteration of the DFS
+* dg: the digraph from which cycles are computed
+* visitor: Informations needed for the cycle computation, contains:
+    * stack: the stack of parent vertices
+    * blocked: tells whether a vertex has already been explored or not
+    * blockedmap: mapping of the blocking / unblocking consequences
+* `vmap`: vector map containing the link from the old to the new nodes of the directed graph
+* startnode = v: optional argument giving the starting node. In the first iteration, 
+the same as v, otherwise it should be passed.
+
+# Returns
+* done: tells whether a circuit has been found in the current exploration.
+"""
+function circuit(v::Int, dg::DiGraph, vis::JohnsonVisitor, vmap::Vector{Int}, startnode = v)
+    done = false
+    push!(vis.stack, v)
+    vis.blocked[v] = true
+    for w in dg.fadjlist[v]
+        if w == startnode
+            produce(vmap[vis.stack])        
+            done = true
+        elseif !vis.blocked[w]
+            circuit(w, dg, vis, vmap, startnode) && (done = true)
+        end
+    end
+    if done
+        unblock!(v, vis.blocked, vis.blockedmap)
+            else
+                for w in dg.fadjlist[v]
+            in(v, vis.blockedmap[w]) || push!(vis.blockedmap[w], v)
```

Comment:
Avoid short circuiting with mutations. See second-to-last bullet at https://github.com/JuliaGraphs/LightGraphs.jl/blob/master/CONTRIBUTING.md.